R/dalys.R

```diff
@@ -0,0 +1,180 @@
+  ## [make remaining life expectancy table] - this will take some time
+  life_table <- create_dalys_life_table(con, touchstone_name, year_min, year_max)
+
+  sets <- unique(dalys_parameters$burden_estimate_set_id)
+
+  dalys_out <- rep(list(NULL),length(sets))
+  for(i in seq_along(sets)){
+    message(sprintf("calculating dalys for burden_estimate_set %s", sets[i]))
+
+    year <- c("(SELECT burden_estimate_set, burden_estimate.country, year, age, burden_outcome, value")
+    year_range <- sprintf("AND year BETWEEN %1s AND %2s", year_min, year_max)
+    sql <- paste("SELECT burden_estimate_set, country, year, age, burden_outcome, code, value as burden, sub_condition, adjusted_weight, duration",
+                 "FROM (SELECT *",
+                 "FROM dalys_parameters",
+                 sprintf("WHERE burden_estimate_set_id = %s", sets[i]),
+                 ") AS parameters",
+                 "LEFT JOIN",
+                 year,
+                 "FROM burden_estimate",
+                 sprintf("WHERE burden_estimate_set = %s", sets[i]),
+                 year_range,
+                 ") AS burdens",
+                 "ON burdens.burden_estimate_set = parameters.burden_estimate_set_id",
+                 "AND burdens.burden_outcome = parameters.burden_outcome_id"
+    )
+    # burden_estimates
+    v <- DBI::dbGetQuery(con, sql)
+    # match dalys parameters
+    v$.code <- paste(v$country, v$year, v$age, sep="-")
+    v <- merge(v, life_table, by= ".code", all.x=TRUE)
+    # make condition-sepcific adjusted duration
+    v$adjusted_duration <- v$duration
+    j <- v$remainning_life_exp < v$duration
+    v$adjusted_duration[j] <- v$remainning_life_exp[j]
+    # calculate dalys
+    v$dalys <- v$burden * v$adjusted_weight * v$adjusted_duration
+    #aggregate to return output in the same structure as burden_estiamte
+    dalys_out[[i]] <- aggregate(dalys ~ burden_estimate_set + country + year + age, data=v, sum, na.rm=TRUE)
+  }
+  # output
+  dat <- do.call(rbind, dalys_out)
+  dat$value <- dat$dalys
+  v <- DBI::dbGetQuery(con, "SELECT * FROM burden_outcome WHERE code = 'dalys'")
+  dat$burden_outcome <- v$id
+  cols <- c("burden_estimate_set", "country", "year", "burden_outcome", "value", "age")
+  dat <- dat[cols]
+}
+
+create_dalys_parameters <- function(con, touchstone_name = "201710gavi", dalys_src, vimc_dalys_only) {
+  ## subsetting those for which VIMC has to calculate DALYs - vimc_dalys = TRUE
+  if(vimc_dalys_only) {
+    dalys_src <- dalys_src[dalys_src$vimc_dalys, ]
+  }
+  ## transform average duration into years; N.B. duration will be compared with remaining life-expectancy
+  ## 1000 years of duration is funny, but helpful
+  i <- unique(dalys_src$period) %in% c("year", "month", "day")
+  if(any(is.na(i))) stop("Unknown period label. Please correct the 'period' column in 'dalys_burden_outcome.csv'.")
+
+  dalys_src$duration <- NA
+  i <- dalys_src$period == "year"
+  dalys_src$duration[i] <- dalys_src$average_duration[i]
```

Comment:
this pattern might be more nicely done as a small function that takes `period` as an argument and returns 1, 1/12 or 1/365.25 then you'd do

```
dalys_src$duration <- dalys_sec$average_duration * duration_weighting(dalys_src$period)
```

Reply:
Done.